src/projectile.cpp

```diff
@@ -92,6 +93,35 @@ void projectile::unset_custom_explosion()
     custom_explosion.reset();
 }
 
+static void foamcrete_build( const tripoint &p )
+{
+    map &here = get_map();
+    const ter_str_id floor = ter_str_id( "t_foamcrete_floor" );
+    const ter_str_id wall = ter_str_id( "t_foamcrete_wall" );
+    const field_type_str_id field_fd_foamcrete( "fd_foamcrete" );
+
+    if( !( wall.is_valid() && floor.is_valid() && field_fd_foamcrete.is_valid() ) ) {
+        debugmsg( "Foamcrete terrains or fields are missing" );
+        return;
+    }
+
+    if( here.has_flag_ter( TFLAG_NO_FLOOR, p ) ) {
+        for( const tripoint &ep : here.points_in_radius( p, 1 ) ) {
+            if( here.has_flag_ter( TFLAG_SUPPORTS_ROOF, ep ) ) {
+                here.ter_set( p, floor );
+                here.add_field( p, field_fd_foamcrete, 1 );
+                return;
+            }
+        }
+        add_msg( m_bad, _( "The foamcrete falls without a wall to anchor against" ) );
+    } else if( here.get_field_intensity( p, field_fd_foamcrete ) >= 2 ) {
+        here.bash( p, 9001, false, true, false );
```

Comment:
This will destroy everything here, including hulks and vehicles?

Reply:
Bash affects only the map, but it destroying vehicles is intended